README.md
# IPFS tools

IPFS tools for [sendgft](https://github.com/sendgft).

This functions as both a library and CLI.


## Library

This exposes a single function - `getIpfsClient()` which is used as follows:

```js
import { getIpfsClient } from '@sendgft/ipfs'

// get client instance
const client = getIpfsClient('http://ipfs-endpoint-url/api/v0')

// upload string
const { cid } = await client.uploadString('test')

// upload string such that it will be available at path /file.txt
const { cid3 } = await client.uploadString('test', '/file.txt')

// upload json
const { cid2 } = await client.uploadJson({ test: true })

// upload json such that it will be available at path /file.txt
const { cid2 } = await client.uploadJson({ test: true }, '/file.txt')
```

## CLI Commands

### daemon

```
yarn ipfs-tools daemon
```

This will start a local IPFS daemon in *test* mode. The endpoints will be:

* API: http://127.0.0.1:5001/api/v0
* HTTP Gateway base URL: http://127.0.0.1:5002/ipfs

_Note: Ensure the [IPFS daemon software](https://ipfs.io/) is installed_.

### upload

```
yarn ipfs-tools upload-defaults --api URL --gateway URL
```

This uploads a file to IPFS and returns the relevant CIDs and URLs for use.

Parameter info:

```
  --api URL       IPFS API endpoint URL. 
  --gateway URL   IPFS gateway base URL. 
```

**Example: upload to local IPFS daemon**

```shell
yarn ipfs-tools upload --file ./test.png --api http://127.0.0.1:5001/api/v0 --gateway http://127.0.0.1:5002/ipfs
```

**Example: upload to Pinata**

```shell
yarn ipfs-tools upload --file ./test.png --api pinata://<pinata api key>:<pinata secret key> --gateway https://ipfs.gft.xyz/ipfs
```


### help

Get help on all available commands.

```
yarn ipfs-tools help
```


## Development

After cloning the repo you must build the code:

```shell
yarn build
```

To auto-rebuild on code changes:

```shell
yarn build-watch
```

## Publishing releases

```shell
yarn release
```

## License

MIT

CHANGELOG.md
# [1.3.0](https://github.com/sendgft/ipfs-tools/compare/v1.2.0...v1.3.0) (2022-05-31)


### Features

* enable folder uploads ([e8b2d5f](https://github.com/sendgft/ipfs-tools/commit/e8b2d5f5064ea2c473764c9a4f3d37bc2bdf7438))

# [1.2.0](https://github.com/sendgft/ipfs-tools/compare/v1.1.0...v1.2.0) (2022-05-17)


### Features

* ensure returned CIDs are always strings ([158b86b](https://github.com/sendgft/ipfs-tools/commit/158b86b6ab66c93dfc60c19fb12811729b954ad5))

# [1.1.0](https://github.com/sendgft/ipfs-tools/compare/v1.0.0...v1.1.0) (2022-05-12)


### Features

* enable directory wrapping on ipfs ([9e00250](https://github.com/sendgft/ipfs-tools/commit/9e00250ad4a35fd6e32c9d1dcfa5d903253b3ae9))

# 1.0.0 (2022-05-11)


### Bug Fixes

* file uploading now works, and added path result param in case we ever need to wrap in a folder ([6e6eec3](https://github.com/sendgft/ipfs-tools/commit/6e6eec3f4c2ca8b240e3a3ce39e19308b5848fe7))
* ipfs versions ([33ce908](https://github.com/sendgft/ipfs-tools/commit/33ce908342f053209a7bdab6e7d0ed49148c6e2b))


### Features

* pass in data folder to upload-defaults ([7e68e74](https://github.com/sendgft/ipfs-tools/commit/7e68e74b59cb16afa34d8851780fdcf69ed40ea3))
* write-default-metadata command ([d4c239c](https://github.com/sendgft/ipfs-tools/commit/d4c239c50382223b09b048b01dd8bc3709733e1d))


src/cli/commands/upload-defaults.ts
import fs from 'fs'
import path from 'path'
import got from 'got'
import { getIpfsClient } from '../..'
import { log, tryCatch } from '../../utils'

export const getMeta = () => ({
  summary: 'Write sendGFT default metadata to IPFS and on-chain contract.',
  params: [
    {
      name: 'folder',
      typeLabel: '{underline path}',
      description: 'The path to the data folder.',
      defaultValue: path.join(process.cwd(), 'data'),
    },
  ],
  options: [
    {
      name: 'api',
      description: 'IPFS API endpoint URL.',
      defaultValue: 'http://127.0.0.1:5001/api/v0',
    },
    {
      name: 'gateway',
      description: 'IPFS gateway base URL.',
      defaultValue: 'http://127.0.0.1:5002/ipfs',
    },
  ]  
})

interface Params {
  folder: string,
  api: string,
  gateway: string,
}

interface UploadDefaultsResult {
  defaultMetadataCid: string,
  openedGftImgCid: string, 
}

export const execute = async ({ folder, api, gateway }: Params): Promise<UploadDefaultsResult> => {
  const GFT_OPENED_SVG = path.join(folder, 'gft-opened.svg')
  const GFT_UNOPENED_SVG = path.join(folder, 'gft-unopened.svg')

  const ipfsClient = getIpfsClient(api)

  const openedGftImgCid: string = await tryCatch('Upload "opened GFT" image to IPFS', async () => {
    const cid = await ipfsClient.uploadFile(GFT_OPENED_SVG)
    log(`Opened GFT image CID: ${cid}`)
    return cid
  })

  const unopenedGftImgCid: string = await tryCatch('Upload "unopened GFT" image to IPFS', async () => {
    const cid = await ipfsClient.uploadFile(GFT_UNOPENED_SVG)
    log(`Unopened GFT image CID: ${cid}`)
    return cid
  })

  if (gateway.substring(-1) !== '/') {
    gateway = `${gateway}/`
  }
  
  const openedGfImgUrl = `${gateway}${openedGftImgCid}`
  const unopenedGfImgUrl = `${gateway}${unopenedGftImgCid}`
  
  // check gateway access
  await tryCatch(`Check that "opened GFT" image exists: ${openedGfImgUrl}`, async () => {
    const ret = await got(openedGfImgUrl)
    const expected = fs.readFileSync(GFT_OPENED_SVG, { encoding: 'utf-8' }).toString()
    if (ret.body !== expected) {
      throw new Error(`Unable to verify "opened GFT" image via gateway: ${openedGfImgUrl}`)
    }
  })
  await tryCatch(`Check that "unopened GFT" image exists: ${unopenedGfImgUrl}`, async () => {
    const ret = await got(unopenedGfImgUrl)
    const expected = fs.readFileSync(GFT_UNOPENED_SVG, { encoding: 'utf-8' }).toString()
    if (ret.body !== expected) {
      throw new Error(`Unable to verify "unopened GFT" image via gateway: ${unopenedGfImgUrl}`)
    }
  })
    
  // upload metadata
  const defaultMetadataCid: string = await tryCatch('Upload metadata to IPFS', async () => {
    const cid = await ipfsClient.uploadJson({
      name: 'Unopened GFT',
      description: 'This is an unopened GFT sent via https://gft.xyz',
      image: unopenedGfImgUrl,
    })
    log(`Unopened GFT metadata CID: ${cid}`)
    return cid
  })

  log(`Default metadata CID: ${defaultMetadataCid}`)
  log(`Opened GFT image CID: ${openedGftImgCid}`)

  return { 
    defaultMetadataCid,
    openedGftImgCid 
  }
}